### PR TITLE
docs:#799 Tiered Storage support for Azure

### LINKED ISSUE #799: 1Q 23.1  Document Tiered Storage support for Azure
### Describe the Issue

Document how self-hosted customers can deploy Redpanda in Azure using tiered storage. 

[RFC Draft](https://docs.google.com/document/d/1RikSuW07TuSr78BMJGJG9HHXOh5XeQ6xP_Ku0gN-JPU)
Redpanda Slack channel: #azure-tiered-storage https://redpandadata.slack.com/archives/C0407G6JBFU

### Who is this for and what problem does it solve?

For Azure cloud customers who want to use shadow indexing. The plan is to mitigate that currently only s3 is supported, with a minio gateway, but as the minio project has now removed that functionality, this is not longer tenable. So Redpanda plans to provide (natively) azure storage support.

### What does success look like for the customer?

They are able to use azure blob storage for shadow indexing - this is what needs to be documented: 
- Prerequisites?
- Configuration steps
- Examples/Use cases (coordinate with PM and customer success what they should be)

### Questions for PM/Eng

- How do users authenticate with Azure blob storage in their environment? Using Managed Identities for VMs, plaintext secrets in Key Vault, something else?

- On the AWS side we have IAM roles for this, so looking to work out what the preferred Azure equivalent is.

**SMEs**:
https://github.com/michaelandrepearce
@jcsp (Engineering)
@Lazin 
@patrickangeles (Customer Success)
@mattschumpert (PM)

### PR DESCRIPTION
**Pull request by micheleRP**
resolves #799 

preview

### CODE REVIEW

**Comment by micheleRP** [comment]
@VladLazar, @Lazin, @bpraseed, @mattschumpert: I realize this could change, but please review the draft doc for steps to [configure access to Azure](https://deploy-preview-939--redpanda-documentation.netlify.app/docs/platform/data-management/tiered-storage/#configure-cloud-storage). I've also added [infrastructure metrics](https://deploy-preview-939--redpanda-documentation.netlify.app/docs/platform/cluster-administration/monitoring/#infrastructure-level-metrics). Please correct anything here and keep me informed of changes as you implement. Thank you!

**Comment by micheleRP** [comment]
@VladLazar, @Lazin, @bpraseed, @mattschumpert: Could one of you please help me understand if we need to document IAM prereqs for Azure, similar to what we have for S3 and GCS [here](https://docs.redpanda.com/docs/platform/security/iam-roles/)? Or do we assume Azure users use the default AD authorization and point [here](https://learn.microsoft.com/en-us/azure/storage/blobs/authorize-data-operations-portal#default-to-azure-ad-authorization-in-the-azure-portal)?

**Comment by VladLazar** [comment]
> @VladLazar, @Lazin, @bpraseed, @mattschumpert: Could one of you please help me understand if we need to document IAM prereqs for Azure, similar to what we have for S3 and GCS [here](https://docs.redpanda.com/docs/platform/security/iam-roles/)? Or do we assume Azure users use the default AD authorization and point [here](https://learn.microsoft.com/en-us/azure/storage/blobs/authorize-data-operations-portal#default-to-azure-ad-authorization-in-the-azure-portal)?

Hey @micheleRP! The plan is to support three auth methods for Azure:
1. Shared Key: user provides a secret that use to sign each request
2. Active Directory: user provides a secret that we exchange for a token and use in each request
3. Active Directory with Managed Identities: user does not provide any secret; the user pins the permissions to the VM and Redpanda fetches them at run-time (see [how-to](https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token) here)

I believe your question is related to (3). Managed Identities work at a higher level than the IAM functionality provided by AWS and GCS. Azure has predefined roles which specify the operations that are allowed on a certain resource. The flow is to create a Managed Identity for a VM and then asign a role (aka RBAC role to it); see [here](https://learn.microsoft.com/en-us/azure/storage/blobs/authorize-managed-identity?source=recommendations). Redpanda needs to perform the same operations in Azure as in S3, so I think the most restrictive role that we can use is ["Storage Account Contributor"](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#storage-account-contributor) applied to the container.

In short, yes, we need to document this flow too.

File: docs/platform/data-management/data-archiving.mdx [hunk 1]
```diff
@@ -16,7 +16,7 @@ Data archiving is one use case for Tiered Storage. With data archiving, you can
 In order to read data from cloud storage, you need to enable remote read. When remote read is disabled, data can only be read from local storage. 
 
 :::note
-Redpanda supports data archiving to Amazon S3 and Google Cloud Storage. 
+Redpanda supports data archiving to Amazon S3, Google Cloud Storage, and Azure Blog Storage.
```
Comment (VladLazar) [inline docs/platform/data-management/data-archiving.mdx:19]:
```suggestion
Redpanda supports data archiving to Amazon S3, Google Cloud Storage, and Azure Blob Storage. 
```

File: docs/platform/data-management/tiered-storage.mdx [hunk 2]
```diff
@@ -98,17 +98,57 @@ To instead configure access to Google Cloud Storage with access keys:
...
+   ```properties
+   cloud_storage_enabled: true
+   cloud_storage_azure_shared_key: ***
+   cloud_storage_azure_client_id: ***
+   cloud_storage_azure_client_secret: ***
+   cloud_storage_region: <eu-north-1>
+   cloud_storage_azure_container: <redpanda_container>
+   ```
+   
+To configure access to an Azure Blob Storage container with Active Directory instead of shared keys:
```
Comment (VladLazar) [inline docs/platform/data-management/tiered-storage.mdx:130]: The plan is to support Shared Key Auth as a hard commit for 23.1. The Active Directory implementation will follow up after we have a reasonable level of confidence in Tiered Storage working well with ABS. I suggest we use the same approach for the docs: only include shared key auth and add Active Directory when it gets merged into Redpanda. AD user configuration seems non-trivial and I'm not sure we can do a good job of documenting it before we have a working implementation for it.

File: docs/platform/data-management/tiered-storage.mdx [hunk 3]
```diff
@@ -98,17 +98,57 @@ To instead configure access to Google Cloud Storage with access keys:
...
+3. Run the `rpk cluster config edit` command, then edit the following required properties:
+
+   ```properties
+   cloud_storage_enabled: true
+   cloud_storage_azure_shared_key: ***
+   cloud_storage_azure_client_id: ***
+   cloud_storage_azure_client_secret: ***
+   cloud_storage_region: <eu-north-1>
+   cloud_storage_azure_container: <redpanda_container>
+   ```
```
Comment (VladLazar) [inline docs/platform/data-management/tiered-storage.mdx:128]: The following cluster configs are required for using ABS with Shared Key Auth: `cloud_storage_enabled`, `cloud_storage_azure_storage_account`, `cloud_storage_azure_container` and `cloud_storage_azure_shared_key`.

`cloud_storage_azure_client_id` and `cloud_storage_azure_client_secret` will be used for one flavour of AD Auth.

 I've included an example Redpanda config that I've used for testing below:
```
redpanda:
  node_id: 0
  internal_topic_replication_factor: 1
  data_directory: "/home/vlad/Work/tmp/data"

  cloud_storage_enabled: true
  cloud_storage_enable_remote_read: true
  cloud_storage_enable_remote_write: true

  cloud_storage_azure_storage_account: vladstorageaccount123
  cloud_storage_azure_container: test-container
  cloud_storage_azure_shared_key: <omitted>
```

File: docs/platform/data-management/tiered-storage.mdx [hunk 4]
```diff
@@ -246,35 +286,35 @@ Starting in release 22.3, when you delete a topic in Redpanda, the data is also
...
 
 ### Reconciliation 
 
-Reconciliation is a Redpanda process that runs periodically on every node. It monitors partitions and decides which partitions are uploaded on each Redpanda node to guarantee that the data is uploaded only once. It also balances the workload evenly between the nodes.
+Reconciliation is a Redpanda process that runs periodically on every node. It monitors partitions and decides which partitions are uploaded on each Redpanda node to guarantee that data is uploaded only once. It also balances the workload evenly between nodes.
 
 When leadership rebalancing occurs, the reconciliation process might stop uploads on one node and start them on another node. 
 
-The reconciliation process is configured with the following property. Edit the property with the `rpk cluster config edit` command:
+The reconciliation process is configured with the `cloud_storage_reconciliation_interval_ms` property. Edit the property with the `rpk cluster config edit` command.
```
Comment (VladLazar) [inline docs/platform/data-management/tiered-storage.mdx:299]: This configuration property has recently been deprecated and doesn't do anything now. In brief, the reconciliation step was removed altogether, and now we react on leadership changes instead. So perhaps, this section should be updated to say that Redpanda reacts to leadership changes, which may cause it to start uploading to cloud storage from a different node.

File: docs/platform/data-management/tiered-storage.mdx [hunk 5]
```diff
@@ -98,17 +98,57 @@ To instead configure access to Google Cloud Storage with access keys:
...
+
+To configure access to Azure Blob Storage with shared keys:
+
+1. Grant a user the following permissions to read and create block blobs on the container to be used with the cluster (or on all containers): `Get Blob`, `Delete Blob`, `Put Blob`.
+2. For shared key authorization, copy the `cloud_storage_azure_shared_key` cluster property. For the client ID and secret for Active Directory authorization, copy `cloud_storage_azure_client_id` and `cloud_storage_azure_client_secret`.
+3. Run the `rpk cluster config edit` command, then edit the following required properties:
+
+   ```properties
+   cloud_storage_enabled: true
+   cloud_storage_azure_shared_key: ***
```
Comment (VladLazar) [inline docs/platform/data-management/tiered-storage.mdx:123]: Do we want to add some indication of how to get the shared key from Azure or do we assume the user will figure it out? The flow is:
1. Log into Azure portal
2. Navigate to the storage account which contains the container you wish for Redpanda to use
3. Select Access Keys in the "Security + networking" tab
4. Retrieve one of the keys

File: docs/platform/data-management/tiered-storage.mdx [hunk 6]
```diff
@@ -398,7 +434,7 @@ Under normal circumstances, you should not need to configure the following prope
...
 | `cloud_storage_upload_ctrl_min_shares` | The minimum number of I/O and CPU shares that the remote write process can use. Default is 100. |
 | `cloud_storage_upload_ctrl_max_shares` | The maximum number of I/O and CPU shares that the remote write process can use. Default is 1000. |
-| `cloud_storage_reconciliation_interval_ms` | Sets the interval, in milliseconds, used to reconcile partitions that need to be uploaded. A long reconciliation interval can result in a delayed reaction to topic creation, topic deletion, or leadership rebalancing events. A short reconciliation interval guarantees that new partitions are picked up quickly, but the process uses more resources. Default is 10000 ms. |
+| `cloud_storage_reconciliation_interval_ms` | The interval, in milliseconds, to reconcile partitions that need to be uploaded. A long reconciliation interval can result in a delayed reaction to topic creation, topic deletion, or leadership rebalancing events. A short reconciliation interval guarantees that new partitions are picked up quickly, but the process uses more resources. Default is 10000 ms. |
```
Comment (VladLazar) [inline docs/platform/data-management/tiered-storage.mdx:437]: This needs to be marked as deprecated. See my previous comment.

File: docs/platform/cluster-administration/monitoring.mdx [hunk 7]
```diff
@@ -148,6 +148,10 @@ Kafka producer rate
 | redpanda_storage_disk_free_space_alert | Status of low storage space alert: 0-OK, 1-low space 2-degraded | none | gauge |
 | redpanda_storage_disk_total_bytes | Total size of attached storage, in bytes. | none | counter |
 | redpanda_uptime_seconds_total | Total CPU busy time in seconds. | none | gauge |
+| redpanda_cloud_client_not_found | Number of requests that came back empty from cloud storage. | --- | --- |
+| redpanda_cloud_client_throttled | Number of client requests that were throttled. | --- | --- |
+| redpanda_cloud_storage_upload_backoff | Number of upload requests that backed off. | --- | --- |
+| redpanda_cloud_storage_download_backoff | Number of download requests that backed off. | --- | --- |
```
Comment (VladLazar) [inline docs/platform/cluster-administration/monitoring.mdx:154]: I also included two metrics which track the total number of uploads and downloads:
`redpanda_cloud_client_uploads` (type counter) and `redpanda_cloud_client_downloads` (type counter).

Note that the labels for these `cloud_client` metrics are different based on whether we are running againts ABS or S3. For S3 we have: `endpoint` (the cloud storage endpoint we are communicating with) and `region` (S3 region being used), while for ABS we have `endpoint` and `storage_account`.

**Review by VladLazar — COMMENTED** [review]
Thanks for working on this, Michelle. I suggest we do AD auth documentation as a follow-up as mentioned in the comments.

File: docs/platform/data-management/tiered-storage.mdx (hunk 6 above)
Comment (micheleRP) [reply docs/platform/data-management/tiered-storage.mdx]: Thanks, @VladLazar. I've marked it as deprecated in this table (and also moved to the bottom of the list), and I did the same for the table on the Tunable Properties page.

File: docs/platform/data-management/tiered-storage.mdx (hunk 5 above)
Comment (micheleRP) [reply docs/platform/data-management/tiered-storage.mdx]: Thanks @VladLazar! Please let me know if the updated section is OK. Also, please confirm that step 1 is required (grant a user permissions for Get Blob, Delete Blob, Put Blob).

File: docs/platform/cluster-administration/monitoring.mdx (hunk 7 above)
Comment (micheleRP) [reply docs/platform/cluster-administration/monitoring.mdx:154]: Thanks, @VladLazar! Please confirm if these look correct/complete!

File: docs/platform/data-management/tiered-storage.mdx (hunk 5 above)
Comment (VladLazar) [reply docs/platform/data-management/tiered-storage.mdx]: As far as I can tell, permissions aren't a thing when using shared key auth, the client gets all perms by default.

File: docs/platform/cluster-administration/monitoring.mdx (hunk 7 above)
Comment (VladLazar) [reply docs/platform/cluster-administration/monitoring.mdx:154]: The labels also differ (in the same way as described in the original comment) for `redpanda_cloud_client_not_found` and `redpanad_cloud_client_throttled`. I've not yet implemented `redpanda_cloud_storage_upload_backoff` and `redpanda_cloud_storage_download_backoff`, but the labels will work the same way.

Also, looking at this again `throttled` and `backoff` have the same semantics, but use different names. Should we pick one and stick with it?

Comment (micheleRP) [reply docs/platform/cluster-administration/monitoring.mdx]: Thanks @VladLazar. I asked @shorgi his opinion. We have precedents for both (max_kafka_throttle_delay_ms, cloud_storage_initial_backoff_ms, ...). He slightly prefers backoff.

**Comment by micheleRP** [comment]
Hi @VladLazar, @Lazin, @mattschumpert: I asked Ben to update the original diagram this page used, so that it wasn't specific to S3. Please see the updated diagram and let me know if you want any changes to it: https://deploy-preview-939--redpanda-documentation.netlify.app/docs/platform/data-management/tiered-storage/

**Comment by mattschumpert** [comment]
@micheleRP I would put the Redpanda node as the top layer and Cloud Storage as the bottom layer.  It's quite odd to think of it as it is now (storage is typically the bottom tier of the cake). Other than that I don't see any major issue.

File: docs/platform/data-management/tiered-storage.mdx [hunk 8]
```diff
@@ -14,47 +14,47 @@ import TabItem from '@theme/TabItem';
...
 :::
 
-Tiered Storage helps to lower storage costs by offloading log segments to cloud storage. You can specify the amount of local storage you want to retain in local storage. You don't need to specify which log segments you want to move because Redpanda moves them automatically based on cluster-level configuration properties. Redpanda Tiered Storage indexes where data is offloaded, so it can retrieve the data when you need it.
+Tiered Storage helps to lower storage costs by offloading log segments to cloud storage. You can specify the amount of storage you want to retain in local storage. You don't need to specify which log segments you want to move, because Redpanda moves them automatically based on cluster-level configuration properties. Tiered Storage indexes where data is offloaded, so it can retrieve the data when you need it.
 
-Redpanda supports Tiered Storage with Amazon S3 and Google Cloud Storage. 
+Redpanda supports Tiered Storage with Amazon S3, Google Cloud Storage (GCS), and Microsoft Azure Blob Storage (ABS).
```
Comment (Lazin) [inline docs/platform/data-management/tiered-storage.mdx:19]: Maybe it will be better to say `Redpanda natively supports ...`? Many systems are claiming that they support ABS but this support is not first party and requires some extra complexity. But redpanda has native support and communicates with ABS directly. Since people will mostly learn about our product through docs it will be nice to send this message.

File: docs/platform/cluster-administration/monitoring.mdx (hunk 7 above)
Comment (Feediver1) [reply docs/platform/cluster-administration/monitoring.mdx]: @VladLazar Can you tell us when these labels will be finalized? Seems like this is lingering a bit. Thx.

Comment (VladLazar) [reply docs/platform/cluster-administration/monitoring.mdx]: Let's go for `backoff` everywhere then (i.e. change `redpanda_cloud_client_throttled` to `redpanda_cloud_client_backoff`. Which labels aren't finalised? They've not changed since my initial comment.

**Comment by micheleRP** [comment]
This PR has been replaced by #1057
